core/src/main/java/org/pentaho/di/core/hash/ByteArrayHashIndex.java

```diff
@@ -0,0 +1,268 @@
+      }
+    }
+
+    // OK, now that we know where to put the entry, insert it...
+    //
+    index[indexPointer] = new ByteArrayHashIndexEntry( hashCode, key, value, index[indexPointer] );
+
+    // Don't forget to link to the previous check entry if there was any...
+    //
+    if ( previousCheck != null ) {
+      previousCheck.nextEntry = index[indexPointer];
+    }
+
+    // If required, resize the table...
+    //
+    resize();
+  }
+
+  private final void resize() {
+    // Increase the size of the index...
+    //
+    size++;
+
+    // See if we've reached our resize threshold...
+    //
+    if ( size >= resizeThresHold ) {
+
+      ByteArrayHashIndexEntry[] oldIndex = index;
+
+      // Double the size to keep the size of the index a factor of 2...
+      // Allocate the new array...
+      //
+      int newSize = 2 * index.length;
+
+      ByteArrayHashIndexEntry[] newIndex = new ByteArrayHashIndexEntry[newSize];
+
+      // Loop over the old index and re-distribute the entries
+      // We want to make sure that the calculation
+      // entry.hashCode & ( size - 1)
+      // ends up in the right location after re-sizing...
+      //
+      for ( int i = 0; i < oldIndex.length; i++ ) {
+        ByteArrayHashIndexEntry entry = oldIndex[i];
+        if ( entry != null ) {
+          oldIndex[i] = null;
+          entry.nextEntry = null; // we assume there is plenty of room in the new index...
+
+          // Make sure we follow all the linked entries...
+          // TODO This is a lot of extra work, see how we can avoid it!
+          //
+          int newIndexPointer = entry.hashCode & ( newSize - 1 );
+
+          // Make sure on this new index pointer, we have room to put the entry
+          //
+          ByteArrayHashIndexEntry check = newIndex[newIndexPointer];
+          if ( check == null ) {
+            // Yes, plenty of room
+            //
+            newIndex[newIndexPointer] = check;
+          } else {
+            // No, we need to look for a nice spot to put the hash entry...
+            //
+            ByteArrayHashIndexEntry previousCheck = null;
+            while ( check != null ) {
+              previousCheck = check;
+              check = check.nextEntry;
+            }
+            while ( newIndex[newIndexPointer] != null ) {
+              newIndexPointer++;
+              if ( newIndexPointer >= newSize ) {
+                newIndexPointer = 0;
+              }
+            }
+            // OK, now that we have a nice spot to put the entry, link the previous check entry to this one...
+            //
+            previousCheck.nextEntry = entry;
+            newIndex[newIndexPointer] = entry;
+          }
+          newIndex[newIndexPointer] = entry;
+        }
+      }
+
+      // Replace the old index with the new one we just created...
+      //
+      index = newIndex;
+
+      // Also change the resize threshold...
+      //
+      resizeThresHold = (int) ( newSize * STANDARD_LOAD_FACTOR );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).